dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/persistence/file/batch/BatchFileHandler.kt

```diff
@@ -126,51 +151,97 @@ internal class BatchFileHandler(
     @Throws(IOException::class)
     @Suppress("UnsafeThirdPartyFunctionCall") // Called within a try/catch block
     private fun readFileData(
-        file: File,
-        prefix: ByteArray,
-        suffix: ByteArray
-    ): ByteArray {
+        file: File
+    ): List<ByteArray> {
         val inputLength = file.lengthSafe().toInt()
-        val outputLength = inputLength + prefix.size + suffix.size
-        val result = ByteArray(outputLength)
 
-        // Copy prefix
-        prefix.copyTo(0, result, 0, prefix.size)
+        val result = mutableListOf<ByteArray>()
 
         // Read file iteratively
-        var offset = prefix.size
         var remaining = inputLength
-        file.inputStream().use {
+        file.inputStream().buffered().use {
             while (remaining > 0) {
-                val read = it.read(result, offset, remaining)
-                if (read < 0) break
-                offset += read
+                val (meta, headerSize) = readEventHeader(it) ?: break
+
+                val eventBytes = ByteArray(meta.eventSize)
+                val readEventSize = it.read(eventBytes, 0, meta.eventSize)
+
+                if (!checkReadSizeExpected(meta.eventSize, readEventSize, "read event")) {
+                    break
+                }
+
+                result.add(eventBytes)
+                val read = headerSize + readEventSize
                 remaining -= read
             }
         }
 
-        // Copy suffix
-        suffix.copyTo(0, result, offset, suffix.size)
-        offset += suffix.size
-
-        return if (result.size == offset) {
-            result
-        } else {
-            result.copyOf(offset)
+        if (remaining != 0) {
+            val message = WARNING_NOT_ALL_DATA_READ.format(Locale.US, file.path)
```

Comment:
Good question. I was thinking about it and decided to keep it to see if this will happen in our nightly tests, just in case.

Not sure if it may happen at all with flash memory (maybe only when the system decides to clean up cache folder while we read the file?), but since we are using the binary format now, it is better to keep an eye on the file integrity to be able to catch possible issues.